src/shared/sleep-config.c

```diff
@@ -526,6 +528,71 @@ int get_total_suspend_interval(Hashmap *last_capacity, usec_t *ret) {
         return 0;
 }
 
+/* Return true if all batteries have acpi_btp support */
+int battery_trip_point_alarm_exists(void) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *dev;
+        int r, battery_count = 0, battery_alarm_count =0;
+
+        r = battery_enumerator_new(&e);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to initialize battery enumerator: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                const char *s;
+                int battery_alarm;
+
+                r = sd_device_get_sysattr_value(dev, "alarm", &s);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to read battery name, ignoring: %m");
+                        continue;
+                }
+
+                r = safe_atoi(s, &battery_alarm);
+                if (r < 0) {
+                        log_device_debug_errno(dev, r, "Failed to parse battery alarm, ignoring: %m");
+                        continue;
+                }
+
+                battery_count++;
+                if (battery_alarm > 0)
+                        battery_alarm_count++;
+        }
+
+        if (battery_count > 0 && battery_count == battery_alarm_count)
+                return 1;
+        return 0;
+}
+
+/* Return true if wakeup type is APM timer */
+int check_wakeup_type(int *ret) {
+        _cleanup_free_ char *s = NULL;
+        size_t readsize;
+        int wakeup_type_bit, r;
+
+        assert(ret);
+
+        /* implementation via dmi/entries */
+        r = read_full_virtual_file(SYS_ENTRY_RAW_FILE_TYPE1, &s, &readsize);
+        if (r < 0)
+                return log_debug_errno(r, "Unable to read %s, ignoring: %m", SYS_ENTRY_RAW_FILE_TYPE1);
+
+        if (readsize < 25 || s[1] < 25)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Only read %zu bytes from %s (expected 24)", readsize, SYS_ENTRY_RAW_FILE_TYPE1);
+
+        wakeup_type_bit = s[24];
+        /* 0 is Reserved and 8 is AC Power Restored. As per table 12 in
+         * https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.4.0.pdf */
+        if (wakeup_type_bit > 0 && wakeup_type_bit < 9) {
+                log_debug("DMI BIOS System Information indicates wakeup type bit is set in range 1-8");
+                *ret = wakeup_type_bit;
```

Comment:
There's no need to expose the byte to the caller. Simply return 1 if it's `3`, 0 if it's not, and negative in case of error or out of the expected range